### PR TITLE
implement inspect-buildpack command

### LINKED ISSUE #642: Pack inspect-buildpack command
### Description
It would be great if `pack` supported a `inspect-buildpack` command. Now that more projects such as Paketo are starting to ship buildpackage artifacts both as `.cnb` artifacts and to a registry, it would be helpful for users to be able to see what's inside of the buildpack. 

### Proposed solution
Create a `pack inspect-buildpack` command. I'm sure we could display a lot of useful buildpack metadata here, but I could see at the minimum it would be great to see the following in the output:
- Buildpack ID
- Buildpack Name
- Buildpack Homepage
- Buildpack Version
- Buildpack Dependencies/Versions

Example using https://github.com/paketo-buildpacks/go/blob/master/buildpack.toml (assuming that the package.toml is similar)
``` 
pack inspect-buildpack go-buildpackage.cnb
.........
ID: "paketo-buildpacks/go"
Name: "Go Buildpack"
Version: "1.0"
Homepage: "https://github.com/paketo-buildpacks/go"

Dependencies:
   ID: "paketo-buildpacks/dep"
   Version: "0.0.109"
   ..........
   ID: "paketo-buildpacks/mod"
   Version: "0.0.112"
```

### Describe alternatives you've considered
Without this command, a user can only see more granular buildpackage metadata by running `inspect-builder` and visiting the homepage URL.

### PR DESCRIPTION
**Pull request by dwillist**
Signed-off-by: dwillist <dthornton@vmware.com>

## Summary
Implements an `inspect-buildpack` command that will allow users to inspect a buildpackage. 

## Output
```
Inspecting buildpack: gcr.io/paketo-buildpacks/nodejs

REMOTE:

Stacks:
  ID: org.cloudfoundry.stacks.cflinuxfs3
    Mixins:
      (none)
  ID: io.buildpacks.stacks.bionic
    Mixins:
      (none)

Buildpacks:
  ID                                    VERSION        HOMEPAGE
  paketo-buildpacks/node-engine         0.0.262        https://github.com/paketo-buildpacks/node-engine
  paketo-buildpacks/node-start          0.0.3
  paketo-buildpacks/nodejs              0.0.6          https://github.com/paketo-buildpacks/nodejs
  paketo-buildpacks/npm                 0.1.80         https://github.com/paketo-buildpacks/npm
  paketo-buildpacks/yarn-install        0.1.87         https://github.com/paketo-buildpacks/yarn-install

Detection Order:
 └ Group #1:
    └ paketo-buildpacks/nodejs@0.0.6
       ├ Group #1:
       │  ├ paketo-buildpacks/node-engine@0.0.262
       │  └ paketo-buildpacks/yarn-install@0.1.87
       ├ Group #2:
       │  ├ paketo-buildpacks/node-engine@0.0.262
       │  └ paketo-buildpacks/npm@0.1.80
       └ Group #3:
          ├ paketo-buildpacks/node-engine@0.0.262
          └ paketo-buildpacks/node-start@0.0.3


LOCAL:

Stacks:
  ID: org.cloudfoundry.stacks.cflinuxfs3
    Mixins:
      (none)
  ID: io.buildpacks.stacks.bionic
    Mixins:
      (none)

Buildpacks:
  ID                                    VERSION        HOMEPAGE
  paketo-buildpacks/node-engine         0.0.262        https://github.com/paketo-buildpacks/node-engine
  paketo-buildpacks/node-start          0.0.3
  paketo-buildpacks/nodejs              0.0.6          https://github.com/paketo-buildpacks/nodejs
  paketo-buildpacks/npm                 0.1.80         https://github.com/paketo-buildpacks/npm
  paketo-buildpacks/yarn-install        0.1.87         https://github.com/paketo-buildpacks/yarn-install

Detection Order:
 └ Group #1:
    └ paketo-buildpacks/nodejs@0.0.6
       ├ Group #1:
       │  ├ paketo-buildpacks/node-engine@0.0.262
       │  └ paketo-buildpacks/yarn-install@0.1.87
       ├ Group #2:
       │  ├ paketo-buildpacks/node-engine@0.0.262
       │  └ paketo-buildpacks/npm@0.1.80
       └ Group #3:
          ├ paketo-buildpacks/node-engine@0.0.262
          └ paketo-buildpacks/node-start@0.0.3

```

## Documentation
Library methods should be documented, this does not require any changes to the `docs` pages. 

- Should this change be documented?
    - [x] Yes

Resolves #642

### CODE REVIEW

**Comment by dfreilich** [comment]
This may relate to #597, in terms of whether it should be `buildpack inspect`, or `inspect buildpack`, etc.

File: inspect_buildpack.go [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+		return buildpackage.Metadata{}, dist.BuildpackLayers{}, errors.Wrapf(err, "invalid registry '%s'", name)
+	}
+
+	registryBp, err := registryCache.LocateBuildpack(name)
+	if err != nil {
+		return buildpackage.Metadata{}, dist.BuildpackLayers{}, errors.Wrapf(err, "unable to find %s in registry: %q", style.Symbol(name), err)
+	}
+	return metadataFromImage(client, registryBp.Address, false)
+}
+func metadataFromArchive(downloader Downloader, path string) (buildpackMd buildpackage.Metadata, layersMd dist.BuildpackLayers, err error) {
```
Comment (dfreilich) [inline inspect_buildpack.go:85]: nit: add line

File: inspect_buildpack.go [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+	var buildpackMd buildpackage.Metadata
+
+	switch locatorType {
+	case buildpack.URILocator:
+		buildpackMd, layersMd, err = metadataFromArchive(c.downloader, opts.BuildpackName)
+	case buildpack.RegistryLocator:
+		buildpackMd, layersMd, err = metadataFromRegistry(c, opts.BuildpackName, opts.Registry)
+	case buildpack.PackageLocator:
+		buildpackMd, layersMd, err = metadataFromImage(c, opts.BuildpackName, opts.Daemon)
+	default:
```
Comment (dfreilich) [inline inspect_buildpack.go:57]: nit: order it `RegistryLocator`, `PackageLocator`, and then `URILocator`, so that you can then order the funcs in order of external and internal usage (`metadataFromRegistry`, `metadataFromImage`, `metadataFromArchive`)

File: inspect_buildpack.go [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+
+type ImgWrapper struct {
+	v1.ImageConfig
+}
+
+func (iw ImgWrapper) Label(name string) (string, error) {
+	return iw.Labels[name], nil
+}
+
+func (c *Client) InspectBuildpack(opts InspectBuildpackOptions) (*BuildpackInfo, error) {
```
Comment (dfreilich) [inline inspect_buildpack.go:42]: I love how this func is decomposed really nicely. Beautiful job

File: inspect_buildpack.go [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+	return buildpackMd, layersMd, nil
+}
+
+func extractOrder(buildpackMd buildpackage.Metadata) dist.Order {
+	return dist.Order{
+		{
+			Group: []dist.BuildpackRef{
+				{
+					BuildpackInfo: buildpackMd.BuildpackInfo,
+					Optional:      false,
```
Comment (dfreilich) [inline inspect_buildpack.go:132]: Is optional meaningful at this point? Do we even need to initialize it?

File: inspect_buildpack_test.go [hunk 5]
```diff
@@ -0,0 +1,490 @@
...
+					"example.com/some/package@sha256:8c27fe111c11b722081701dfed3bd55e039b9ce92865473cf4cdfa918071c566",
+					false,
+					config.PullNever).Return(buildpackImage, nil)
+			})
+
+			it.After(func() {
+				h.AssertNil(t, os.Unsetenv("PACK_HOME"))
+			})
+
+			it("succeeds", func() {
```
Comment (dfreilich) [inline inspect_buildpack_test.go:373]: nit: Could you also add a test for the case where it errors returning from the registry?

File: inspect_buildpack_test.go [hunk 6]
```diff
@@ -0,0 +1,490 @@
...
+			it.Before(func() {
+				expectedInfo.Location = buildpack.URILocator
+
+				pack.WithDownloader(mockDownloader)(subject)
+
+				assert := h.NewAssertionManager(t)
+				writeBuildpackArchive(buildpackPath, tmpDir, assert)
+			})
+
+			it("succeeds", func() {
```
Comment (dfreilich) [inline inspect_buildpack_test.go:395]: nit: Could you also add a test for a failed download?

File: inspect_buildpack_test.go [hunk 7]
```diff
@@ -0,0 +1,490 @@
...
+						info, err := subject.InspectBuildpack(inspectOptions)
+						h.AssertNil(t, err)
+
+						h.AssertEq(t, info, expectedInfo)
+					})
+				})
+			}
+		})
+	})
+	when("failure cases", func() {
```
Comment (dfreilich) [inline inspect_buildpack_test.go:435]: 🥳 🥳 🥳

File: internal/commands/inspect_buildpack.go [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+					BuildpackName: args[0],
+					Daemon:        true,
+					Registry:      cfg.DefaultRegistry,
+				},
+			)
+			if err != nil {
+				logger.Errorf("inspecting buildpack '%s': %s", buildpackName, err)
+			}
+
+			prefix := determinePrefix(args[0], daemonInspect.Location, true)
```
Comment (dfreilich) [inline internal/commands/inspect_buildpack.go:82]:
```suggestion
			prefix := determinePrefix(buildpackName, daemonInspect.Location, true)
```

File: internal/commands/inspect_buildpack.go [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+		Use:   "inspect-buildpack <image-name>",
+		Short: "Show information about a buildpack",
+		Args:  cobra.MaximumNArgs(2),
+		RunE: logError(logger, func(cmd *cobra.Command, args []string) error {
+			buildpackName := args[0]
+			logger.Infof("Inspecting buildpack: %s\n", style.Symbol(buildpackName))
+
+			daemonInspect, err := client.InspectBuildpack(
+				pack.InspectBuildpackOptions{
+					BuildpackName: args[0],
```
Comment (dfreilich) [inline internal/commands/inspect_buildpack.go:73]:
```suggestion
					BuildpackName: buildpackName,
```

File: internal/commands/inspect_buildpack.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+			prefix := determinePrefix(args[0], daemonInspect.Location, true)
+			output, err := inspectBuildpackOutput(daemonInspect, prefix, flags.Depth)
+			if err != nil {
+				return err
+			}
+
+			var extraOutput string
+			if daemonInspect.Location == buildpack.PackageLocator {
+				remoteInspect, err := client.InspectBuildpack(pack.InspectBuildpackOptions{
+					BuildpackName: args[0],
```
Comment (dfreilich) [inline internal/commands/inspect_buildpack.go:91]:
```suggestion
					BuildpackName: buildpackName,
```

File: internal/commands/inspect_buildpack.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+					Daemon:        false,
+					Registry:      cfg.DefaultRegistry,
+				})
+				if err != nil {
+					logger.Errorf("inspecting buildpack '%s': %s", buildpackName, err)
+				}
+				extraPrefix := determinePrefix(args[0], remoteInspect.Location, false)
```
Comment (dfreilich) [inline internal/commands/inspect_buildpack.go:98]:
```suggestion
				extraPrefix := determinePrefix(buildpackName, remoteInspect.Location, false)
```

File: internal/commands/inspect_buildpack_test.go [hunk 12]
```diff
@@ -0,0 +1 @@
+package commands
```
Comment (dfreilich) [inline internal/commands/inspect_buildpack_test.go:1]: nit: Add tests? 😆

**Review by dfreilich — COMMENTED** [review]
Really great start! I'd like to see a few more tests on some of the negative code paths, and a test or two on the command, but definitely great work. 

I just wonder what the name of the command should be, given that the renaming rfc is due to be merged tomorrow.

File: internal/commands/inspect_buildpack_test.go [hunk 13]
```diff
@@ -0,0 +1,660 @@
...
+						"urn:cnb:registry:test/buildpack",
+						"REGISTRY IMAGE:",
+						simpleOutputSection)
+
+					assert.TrimmedEq(outBuf.String(), expectedOutput)
+				})
+			})
+		})
+
+		when("a depth flag is passed", func() {})
```
Comment (dfreilich) [inline internal/commands/inspect_buildpack_test.go:546]:
```suggestion
		when("a depth flag is passed", func() {
```

**Review by dfreilich — APPROVED** [review]
Really great work! Especially love the tests, particularly the focus on rigorous sad path tests. 

safety

**Comment by dfreilich** [comment]
# User Acceptance
## Local Image
image

## Remote Image
image

## Local and Remote Image
image

## Verbose
image

## Adjusted Depth
image

## Local Archive (downloaded from [here](https://github.com/paketo-buildpacks/nodejs/releases))
image

## Remote Archive
image

## Registry Image
With this in my ~/.pack/config.toml:
```toml
default-registry="elbandito"
default-registry-url="https://github.com/elbandito/buildpack-registry"
```
image

## Unknown Image
image

File: internal/commands/inspect_buildpack.go [hunk 14]
```diff
@@ -0,0 +1,188 @@
...
+	AddHelpFlag(cmd, "inspect-buildpack")
+	return cmd
+}
+
+func inspectAllBuildpacks(client PackClient, flags InspectBuildpackFlags, options ...pack.InspectBuildpackOptions) (string, error) {
+	buf := bytes.NewBuffer(nil)
+	skipCount := 0
+	for _, option := range options {
+		nextResult, err := client.InspectBuildpack(option)
+		// when do we exit
```
[the remark this answers is not in the document]
Comment (dwillist) [reply internal/commands/inspect_buildpack.go:110]: Hahaha. Ya lemme just fix that reallll quick.

**Review by jromero — CHANGES_REQUESTED** [review]
This looks great! Just a few minor changes that @dfreilich already pointed out.

File: inspect_buildpack_test.go [hunk 15]
```diff
@@ -0,0 +1,604 @@
...
+				})
+				it("returns an error", func() {
+					inspectOptions := pack.InspectBuildpackOptions{
+						BuildpackName: "https://invalid/buildpack",
+						Daemon:        true,
+					}
+
+					_, err := subject.InspectBuildpack(inspectOptions)
+					h.AssertNotNil(t, err)
+					h.AssertTrue(t, !errors.Is(err, image.ErrNotFound))
```
[the remark this answers is not in the document]
Comment (dwillist) [reply inspect_buildpack_test.go:495]: Ya these extra comparisons are just to be explicit about the error 'type'. The `image.ErrNotFound` is used to differentiate the following cases:

- `fetcher.Fetch(...)` call failed to retrieve an image because it does not exist (returns `image.ErrNotFound`, we just tried to fetch an image that does not exist & can probably continue program execution ).
- `fetcher.Fetch(...)` returned some other error when attempting to retrieve the image. (return any other error type, something terrible happened and we should probably exit)

Happy to add string matching as well. But I think we might want these explicit 'type' checks.

**Review by jromero — APPROVED** [review]
